stac/harness/ilp.py

```diff
@@ -0,0 +1,188 @@
+""" ILP decoding """
+from __future__ import print_function
+
+import os
+import re
+import itertools as itr
+from os import path as fp
+import numpy as np
+from tempfile import mkdtemp
+from shutil import rmtree
+from subprocess import call
+
+from attelo.table import UNRELATED
+from attelo.decoding import Decoder
+
+ZPL_TEMPLATE_DIR = fp.join(fp.dirname(__file__), 'ilp')
+
+
+def pos_indexes(dpack):
+    """ Returns indices of EDUs for each pairing """
+    edu_pos = dict((e, i) for i, e in enumerate(dpack.edus))
+    pair_pos = np.vstack((edu_pos[u], edu_pos[v]) for u, v in dpack.pairings)
+
+    return tuple(pair_pos.transpose())
+
+
+def zimpl_dump(dpack, tgt_dir=None, prefix='default', decoded=False):
+    """ Dump classification scores for use in SCIP
+
+    Default behavior is a dump of dpack.attach and dpack.label
+    If decoded is True, dpack.prediction will be converted
+    """
+    n_edus = len(dpack.edus)
+    n_labels = len(dpack.labels)
+    unrelated = dpack.label_number(UNRELATED)
+
+    pair_pos = pos_indexes(dpack)
+
+    tmpdir = mkdtemp() if tgt_dir is None else tgt_dir
+    format_str = '{0:.0f}' if decoded else '{0:.2f}'
+
+    # Attachments
+    att_file = os.path.join(tmpdir, '{0}.attach.dat'.format(prefix))
+    att_mat = np.zeros((n_edus, n_edus), dtype=float)
+    if decoded:
+        att_mat[pair_pos] = (
+            np.array(dpack.graph.prediction != unrelated, dtype=int))
+    else:
+        att_mat[pair_pos] = dpack.graph.attach
+
+    with open(att_file, 'w') as f:
+        f.write('\n'.join(
+                    ':'.join(format_str.format(p)
+                    for p in row)
+                for row in att_mat))
+        f.write('\n')
+
+    # Labels
+    lab_file = os.path.join(tmpdir, '{0}.label.dat'.format(prefix))
+    lab_tsr = np.zeros((n_edus, n_edus, n_labels), dtype=float)
+    if decoded:
+        attached_mask = (dpack.graph.prediction != unrelated)
+        lab_tsr[pair_pos[0][attached_mask],
+                pair_pos[1][attached_mask],
+                dpack.graph.prediction[attached_mask]] = 1
+    else:
+        lab_tsr[pair_pos] = dpack.graph.label
+
+    with open(lab_file, 'w') as f:
+        f.write('\n'.join(
+                    ' '.join(
+                        ':'.join(format_str.format(p)
+                        for p in tube)
+                    for tube in row)
+                for row in lab_tsr))
+        f.write('\n')
+
+    return tmpdir
+
+
+def pretty_data(data):
+    """ Formats a list of lists, with space and linebreaks separation """
+    return '\n'.join(
+                ' '.join(str(e) for e in lis)
+            for lis in data)
+
+def mk_zimpl_input(dpack, data_dir):
+    """ Create ZIMPL input file tuned to the current data directory """
```

Comment:
Could you flesh out the docstring here? More information about parameters and result would be great, plus a brief description of what this is doing.